File: architecture-skills/object-oriented-programming-paradigm/examples/open-closed/open-closed.js
// Open/Closed Principle in JavaScript.
//
// PricingEngine is closed: it iterates registered rules through the
// discount duck contract and never names a concrete rule. FlashSale
// is added later as a new class -- zero edits to the engine.

class PricingEngine {
  constructor() {
    this._rules = [];
  }

  register(rule) {
    this._rules.push(rule);
  }

  total(order) {
    let total = order.subtotal;
    for (const rule of this._rules) {
      const amount = rule.discount(order);
      if (amount > 0) {
        console.log(`  ${rule.name.padStart(6)}: -$${amount.toFixed(2)}`);
      }
      total -= amount;
    }
    return total;
  }
}

// --- Extensions: the engine never mentions these types ---

class BulkDiscount {
  get name() {
    return 'bulk';
  }

  discount(order) {
    return order.itemCount >= 10 ? 0.05 * order.subtotal : 0.0;
  }
}

class MemberDiscount {
  get name() {
    return 'member';
  }

  discount(order) {
    return order.isMember ? 0.10 * order.subtotal : 0.0;
  }
}

// A NEW rule, written later -- nothing above is touched.
class FlashSale {
  constructor(amount, threshold) {
    this._amount = amount;
    this._threshold = threshold;
  }

  get name() {
    return 'flash';
  }

  discount(order) {
    return order.subtotal > this._threshold ? this._amount : 0.0;
  }
}

function main() {
  const order = { subtotal: 120.0, itemCount: 12, isMember: true };

  const engine = new PricingEngine();
  engine.register(new BulkDiscount());
  engine.register(new MemberDiscount());

  console.log(`subtotal: $${order.subtotal.toFixed(2)}`);
  console.log(`total:    $${engine.total(order).toFixed(2)}`);

  console.log('-- registering FlashSale: no engine edits --');
  engine.register(new FlashSale(15.0, 100.0));
  console.log(`total:    $${engine.total(order).toFixed(2)}`);
}

main();

File: architecture-skills/object-oriented-programming-paradigm/examples/open-closed/open-closed.java
// Open/Closed Principle in Java.
//
// PricingEngine is closed: it iterates registered DiscountRule
// implementations and never names a concrete rule. FlashSale is
// added later as a new class -- zero edits to the engine. (In real
// systems the registrations come from DI or ServiceLoader.)
//
// Java identifiers cannot contain hyphens, so the class cannot be named
// after this file; a package-private class keeps the single-file source
// launcher (`java open-closed.java`) working.

import java.util.ArrayList;
import java.util.List;

class OpenClosedDemo {

    record Order(double subtotal, int itemCount, boolean isMember) {}

    /** Extension point: one implementation per business rule. */
    interface DiscountRule {
        String name();
        double discount(Order order);
    }

    // --- Extensions: the engine never mentions these types ---

    static final class BulkDiscount implements DiscountRule {
        public String name() { return "bulk"; }
        public double discount(Order o) {
            return o.itemCount() >= 10 ? 0.05 * o.subtotal() : 0.0;
        }
    }

    static final class MemberDiscount implements DiscountRule {
        public String name() { return "member"; }
        public double discount(Order o) {
            return o.isMember() ? 0.10 * o.subtotal() : 0.0;
        }
    }

    /** Closed core: computes a total over whatever rules it was given. */
    static final class PricingEngine {
        private final List<DiscountRule> rules = new ArrayList<>();

        void register(DiscountRule rule) {
            rules.add(rule);
        }

        double total(Order order) {
            double total = order.subtotal();
            for (DiscountRule rule : rules) {
                double amount = rule.discount(order);
                if (amount > 0.0)
                    System.out.printf("  %6s: -$%.2f%n", rule.name(), amount);
                total -= amount;
            }
            return total;
        }
    }

    /** A NEW rule, written later -- nothing above is touched. */
    static final class FlashSale implements DiscountRule {
        private final double amount;
        private final double threshold;

        FlashSale(double amount, double threshold) {
            this.amount = amount;
            this.threshold = threshold;
        }

        public String name() { return "flash"; }
        public double discount(Order o) {
            return o.subtotal() > threshold ? amount : 0.0;
        }
    }

    public static void main(String[] args) {
        Order order = new Order(120.00, 12, true);

        PricingEngine engine = new PricingEngine();
        engine.register(new BulkDiscount());
        engine.register(new MemberDiscount());

        System.out.printf("subtotal: $%.2f%n", order.subtotal());
        System.out.printf("total:    $%.2f%n", engine.total(order));

        System.out.println("-- registering FlashSale: no engine edits --");
        engine.register(new FlashSale(15.00, 100.00));
        System.out.printf("total:    $%.2f%n", engine.total(order));
    }
}
